pkg/envoy/xds/ack.go

```diff
@@ -93,23 +111,38 @@ type pendingCompletion struct {
 // a string identifier from an Envoy Node identifier.
 func NewAckingResourceMutatorWrapper(mutator ResourceMutator, nodeToID NodeToIDFunc) *AckingResourceMutatorWrapper {
 	return &AckingResourceMutatorWrapper{
-		mutator:  mutator,
-		nodeToID: nodeToID,
+		mutator:            mutator,
+		nodeToID:           nodeToID,
+		pendingCompletions: make(map[*completion.Completion]*pendingCompletion),
 	}
 }
 
 func (m *AckingResourceMutatorWrapper) Upsert(typeURL string, resourceName string, resource proto.Message, nodeIDs []string, completion *completion.Completion) {
 	m.locker.Lock()
 	defer m.locker.Unlock()
 
-	// Always upsert the resource, even if the completion's context was
-	// canceled before we even started, since we have no way to signal whether
-	// the resource is actually upserted.
+	// Caller is expected to use the completion callbacks to get notified when the resource has been acknowledged
+	// If the context is cancelled no such acknowledgement is made
+	if completion.Context().Err() != nil {
```

Comment:
You must do that only _after_ testing that `completion` is not `nil`, which is currently done below.

Reply:
Yep, noticed that when editing for the other comments :-)